gpu4pyscf/grad/uhf.py

```diff
@@ -0,0 +1,369 @@
+    t0 = log.init_timer()
+
+    mo_energy = cupy.asarray(mo_energy)
+    mo_occ = cupy.asarray(mo_occ)
+    mo_coeff = cupy.asarray(mo_coeff)
+    dm0 = mf.make_rdm1(mo_coeff, mo_occ)
+    dme0 = mf_grad.make_rdm1e(mo_energy, mo_coeff, mo_occ)
+    dm0 = tag_array(dm0, mo_coeff=mo_coeff, mo_occ=mo_occ)
+    dm0_sf = dm0[0] + dm0[1]
+    dme0_sf = dme0[0] + dme0[1] 
+    
+    s1 = mf_grad.get_ovlp(mol)
+    # vhf = mf_grad.get_veff(mol, dm0)
+    
+    if atmlst is None:
+        atmlst = range(mol.natm)
+    aoslices = mol.aoslice_by_atom()
+    de = cupy.zeros((len(atmlst),3))
+        
+    def calculate_h1e(h1_gpu, s1_gpu):
+        # (\nabla i | hcore | j) - (\nabla i | j)
+        h1_cpu = mf_grad.get_hcore(mol)
+        s1_cpu = mf_grad.get_ovlp(mol)
+        h1_gpu[:] = cupy.asarray(h1_cpu)
+        s1_gpu[:] = cupy.asarray(s1_cpu)
+        return
+
+    h1 = cupy.empty([3, dm0.shape[1], dm0.shape[2]])
+    s1 = cupy.empty([3, dm0.shape[1], dm0.shape[2]])
+    with lib.call_in_background(calculate_h1e) as calculate_hs:
+        calculate_hs(h1, s1)
+        # (i | \nabla hcore | j)
+        t3 = log.init_timer()
+        dh1e = int3c2e.get_dh1e(mol, dm0_sf)
+
+        t4 = log.timer_debug1("get_dh1e", *t3)
+        if mol.has_ecp():
+            dh1e += rhf_grad.get_dh1e_ecp(mol, dm0_sf)
+        t1 = log.timer_debug1('gradients of h1e', *t0)
+        log.debug('Computing Gradients of NR-HF Coulomb repulsion')
+        dvhf = mf_grad.get_veff(mol, dm0)
+
+        extra_force = cupy.zeros((len(atmlst),3))
+        for k, ia in enumerate(atmlst):
+            extra_force[k] += mf_grad.extra_force(ia, locals())
+
+        t2 = log.timer_debug1('gradients of 2e part', *t1)
+
+    dh = contract('xij,ij->xi', h1, dm0_sf)
+    ds = contract('xij,ij->xi', s1, dme0_sf)
+    delec = 2.0*(dh - ds)
+    delec = cupy.asarray([cupy.sum(delec[:, p0:p1], axis=1) for p0, p1 in aoslices[:,2:]])
+    
+    # dvhf = mf_grad.get_veff(mol, dm0)
+#     for k, ia in enumerate(atmlst):
+#         shl0, shl1, p0, p1 = aoslices[ia]
+# # s1, vhf are \nabla <i|h|j>, the nuclear gradients = -\nabla
+#         de[k] += cupy.einsum('sxij,sij->x', vhf[:,:,p0:p1], dm0[:,p0:p1]) * 2
+    # de += 2*dvhf
+    # de += dh1e + delec + extra_force
+    de = 2.0 * dvhf + dh1e + delec + extra_force
+    
+
+    if(hasattr(mf, 'disp') and mf.disp is not None):
+        g_disp = mf_grad.get_dispersion()
+        mf_grad.grad_disp = g_disp
+        mf_grad.grad_mf = de
+        de += cupy.asarray(g_disp)
+
+    de -= cupy.sum(de, axis=0)/len(atmlst)
+    return de.get()
+
+
+class Gradients(uhf.Gradients):
+    from gpu4pyscf.lib.utils import to_cpu, to_gpu, device
+    
+    grad_elec = grad_elec
+    grad_nuc = rhf_grad.grad_nuc 
+    # get_veff = rhf_grad.get_veff 
+    get_veff =  get_veff
+    get_jk = rhf_grad._get_jk 
+    
+    # def get_jk(self, mol=None, dm=None, hermi=0, omega=None):
+    #     if mol is None: mol = self.mol
```

Comment:
Delete this?

Reply:
OK, I will clean up the codes these days.